File: simulator/domain_conditions.py
from pathlib import Path
from dataclasses import dataclass
from enum import Enum, auto

import tomllib
import numpy as np
from simulator.mesh import Mesh


class ConditionType(Enum):
    """
    First type condtions specifies the values of variable applied at the boundary (AKA Dirichlet)
    Second type condtions specifies the values of the derivative applied at the boundary of the domain. (AKA  Neumann condition)
    """

    FIRST = auto()
    SECOND = auto()
    # TODO: maybe implement in the future the other conditions (Cauchy and Robin) with combination of these two others


# TODO: check if this could be better a jitclass
@dataclass
class Conditions:
    """It holds the conditions (initial or boundary) to be applied to vectors/matrices before solving process"""

    indices: list
    values: list


# TODO: check if this could be better a jitclass
class DomainConditions:
    """
    This holds data related to initial and boundary conditions
    First type condtions here are Dirichlet conditions (prescribed values)
    Second type condtions here are Newmann conditions (prescribed derivative values)
    """

    def __init__(
        self,
        conditions_filepath: Path,
        mesh: Mesh,
        default_initial_values: dict[str, float],
    ):
        with open(conditions_filepath, "rb") as f:
            self.conditions_data = tomllib.load(f)
        self.validate_conditions_data()
        # create the conditions
        self.initial_conditions = {
            (
                condition_data["variable_name"],
                condition_data["condition_type"],
            ): Conditions([], [])
            for condition_data in self.conditions_data["initial"]
        }
        self.boundary_conditions = {
            (
                condition_data["variable_name"],
                condition_data["condition_type"],
            ): Conditions([], [])
            for condition_data in self.conditions_data["boundary"]
        }
        self._setup_user_defined_conditions(mesh)

    def validate_conditions_data(self):
        # TODO: do a logical validation here to ensure :
        # - all group names are valid names
        # - all variables are valid variable names (just alert if not)
        # - valid condition type number (raise by now if is not first type not implemented yet)
        # - alert duplicated conditions and pop from imported data
        # - break if there are two conditions with same (named group + variable + condition type)
        #   message ambiguous or duplicated
        # - for now only allow first type condition (value conition) for initial values
        #   change this for validation process
        pass

    def _setup_user_defined_conditions(self, mesh):
        """
        Setup the set of indices and values for each variable and condition type configured
        in the condition files loaded.
        """
        # setup initial and boundary conditions from the conditions data file
        for node_id in range(mesh.nodes_handler.total_nodes):
            for condition_data in self.conditions_data["initial"]:
                # if the group name is empty in the input file then define it always the condition
                if len(condition_data["group_name"]) == 0:
                    node_in_group = True
                else:
                    group_number = mesh.named_groups[condition_data["group_name"]]
                    node_in_group = (
                        mesh.nodes_handler.nodes[node_id].named_group == group_number
                    )
                if node_in_group:
                    variable_name = condition_data["variable_name"]
                    # only apply first condition type as initial values
                    condition_type = ConditionType.FIRST.value
                    value = condition_data["value"]
                    self.initial_conditions[
                        (variable_name, condition_type)
                    ].indices.append(node_id)
                    self.initial_conditions[
                        (variable_name, condition_type)
                    ].values.append(value)
            for condition_data in self.conditions_data["boundary"]:
                # if the group name is empty in the input file then define it always the condition
                if len(condition_data["group_name"]) == 0:
                    node_in_group = True
                else:
                    group_number = mesh.named_groups[condition_data["group_name"]]
                    node_in_group = (
                        mesh.nodes_handler.nodes[node_id].named_group == group_number
                    )
                if node_in_group:
                    variable_name = condition_data["variable_name"]
                    condition_type = condition_data["condition_type"]
                    value = condition_data["value"]
                    self.boundary_conditions[
                        (variable_name, condition_type)
                    ].indices.append(node_id)
                    self.boundary_conditions[
                        (variable_name, condition_type)
                    ].values.append(value)

    def apply_initial_conditions(self, nodes_handler):
        """Apply user defined initial values for the variables"""
        # apply initial condition values for the nodes
        for variable_condition_type, condition in self.initial_conditions.items():
            variable_name, _ = variable_condition_type
            for index, value in zip(condition.indices, condition.values):
                nodes_handler.nodes[index].variables[variable_name] = value
        # apply boundary values (first type condition type) for the nodes
        for variable_condition_type, condition in self.boundary_conditions.items():
            variable_name, condition_type = variable_condition_type
            if condition_type == ConditionType.FIRST.value:
                for index, value in zip(condition.indices, condition.values):
                    nodes_handler.nodes[index].variables[variable_name] = value

    def get_lhs_with_boundary_condition(self, lhs, variable_name):
        """Return the LHS of equation with boundary conditions applied"""
        lhs_boundary_applied = lhs.copy()
        for index in self.boundary_conditions[
            (variable_name, ConditionType.FIRST.value)
        ].indices:
            lhs_boundary_applied[:, index] = 0.0
            lhs_boundary_applied[index, :] = 0.0
            lhs_boundary_applied[index, index] = 1.0
        lhs_boundary_applied.eliminate_zeros()
        # TODO: check how to apply condition for the other type here
        return lhs_boundary_applied

    def get_rhs_with_boundary_condition(self, lhs, rhs, variable_name):
        """
        Return the RHS of equation with boundary conditions applied
        NOTE: this function expect a one-dimension vector for RHS
        """
        # first condition application
        boundary_conditions = self.boundary_conditions[
            (variable_name, ConditionType.FIRST.value)
        ]
        offset_vector = self._calculate_rhs_offset_values(lhs, variable_name)
        rhs_boundary_applied = rhs + offset_vector
        rhs_boundary_applied[boundary_conditions.indices] = boundary_conditions.values
        # TODO: do the calculations for the other condition_types
        return rhs_boundary_applied

    def _calculate_rhs_offset_values(self, lhs, variable_name: str):
        """Calculate the vector to be added to rhs vector due boundary condition application"""
        boundary_conditions = self.boundary_conditions[
            (variable_name, ConditionType.FIRST.value)
        ]
        total_rows = lhs.shape[0]
        offset = np.zeros((total_rows, 1), dtype=np.float64)
        # accumulate column vectors in sparse matrix with boundary indices
        for column_id, value in zip(
            boundary_conditions.indices, boundary_conditions.values
        ):
            offset -= lhs[:, [column_id]] * value
        # TODO: apply conditions of second type here ...
        # ensure zeros in offset vector in positions where boundary are applied
        # this is needed to not mess values when other t
        offset[boundary_conditions.indices] = 0.0
        return offset.reshape(total_rows)

    def apply_first_type_condition(self, variable_name, variable_values):
        boundary_conditions = self.boundary_conditions[
            (variable_name, ConditionType.FIRST.value)
        ]
        variable_values[boundary_conditions.indices] = boundary_conditions.values

File: simulator/src/domain_conditions.jl
export DomainConditions
export ConditionType


"""
First type condtions specifies the values of variable applied at the boundary (AKA Dirichlet)
Second type condtions specifies the values of the derivative applied at the boundary of the domain. (AKA  Neumann condition)
"""
@enum ConditionType begin
    FIRST = 1
    SECOND = 2
    # TODO: maybe implement in the future the other conditions (Cauchy and Robin) with combination of these two others
end


"""
This struct holds data related to domain conditions
NOTE: initial conditions is already applied by the fronted code i.e. the initial values
are imported in the simulation
"""
struct DomainConditions
    # indices and values of nodes that have domain conditions values
    # the keys are the unknown name and condition type
    indices::Dict{Tuple{String, ConditionType}, Vector{Int64}}
    values::Dict{Tuple{String, ConditionType}, Vector{Float64}}
end


"""Load data for domain conditions needed to the simulation"""
function load_domain_conditions(domain_conditions_groups, domain_conditions_data)
    indices = Dict{Tuple{String, ConditionType}, Vector{Int64}}()
    values = Dict{Tuple{String, ConditionType}, Vector{Float64}}()
    # get boundary conditions
    for condition_data in domain_conditions_data["boundary"]
        group_number = condition_data["group_name"]
        unknown = condition_data["unknown"]
        value = condition_data["value"]
        condition_type = get_condition_type(condition_data["condition_type"])
        curent_group_indices = findall(
            domain_condition_group -> domain_condition_group == group_number, 
            domain_conditions_groups
        )
        indices[(unknown, condition_type)] = curent_group_indices
        values[(unknown, condition_type)] = fill(value, length(curent_group_indices))
    end
    return DomainConditions(indices, values)
end


"""Auxiliary function to convert a type number of condition into the enum values"""
function get_condition_type(condition_type_number)
    if condition_type_number == 1
        return FIRST
    elseif condition_type_number == 2
        return SECOND
    else
        throw("Not implement group number of type $condition_type_number")
    end
end


"""Setup initial boundary values for the unknowns"""
function setup_boundary_values(
    domain_conditions::DomainConditions,
    unknowns_handler::UnknownsHandler,
)
    for unknown in keys(unknowns_handler.values)
        if haskey(domain_conditions, (FIRST, unknown))
            indices = domain_conditions[(FIRST, unknown)].indices
            values = domain_conditions[(FIRST, unknown)].values
            unknowns_handler.values[unknown][indices] = values
            unknowns_handler.old_values[unknown][indices] = values
        end
    end
end



"""Apply boundary conditions to the LHS matrix"""
function apply_domain_conditions_lhs!(
    domain_conditions::DomainConditions, 
    unknown::String,
    lhs::SparseMatrixCSC{Float64, Int64},
)
    #logger.info("applying conditions to LHS")
    # output_manager.write_debug(
    #     f"{self.label}/{unknown}/lhs_condition_applied",
    #     self.lhs_condition_applied[unknown],
    # )
    for index in domain_conditions.indices[(unknown, FIRST::ConditionType)]
        lhs[:, index] .= 0.0
        lhs[index, :] .= 0.0
        lhs[index, index] = 1.0
    end
    # TODO: check how to apply condition for the other type here
end


"""Apply boundary conditions to the RHS vector"""
function apply_domain_conditions_rhs!(
    domain_conditions::DomainConditions, 
    unknown::String,
    assembled_lhs::SparseMatrixCSC{Float64, Int64},
    rhs::Vector{Float64},
)
    #logger.info("applying conditions to RHS")
    # output_manager.write_debug(
    #     f"{self.label}/{unknown}/rhs_condition_applied",
    #     rhs_condition_applied,
    # )
    # first condition application
    indices = domain_conditions.indices[
        (unknown, ConditionType.FIRST.value)
    ]
    values = domain_conditions.values[
        (unknown, ConditionType.FIRST.value)
    ]
    offset_vector = calculate_rhs_offset_values(
        assembled_lhs, indices, values
    )
    rhs = rhs - offset_vector
    rhs[indices] .= values
    # TODO: do the calculations for the other condition_types
end


"""
Calculate the vector to be added to rhs vector due domain condition application.
To zero a column of the LHS matrix the column values must be multiplied by the known
value and added to the RHS vector to give a right offset.
"""
function calculate_rhs_offset_values(
    assembled_lhs::SparseMatrixCSC{Float64, Int64},
    indices::Vector{Int64},
    values::Vector{Float64}
)
    # offset vector must be same size of the amount of LHS rows
    offset = zeros(assembled_lhs.m)
    # accumulate column vectors in sparse matrix with boundary indices
    for (column_id, value) in zip(indices, values)
        offset += collect(assembled_lhs[:, column_id] * value)
    end
    # ensure zeros in offset vector in positions where boundary are applied
    offset[indices] .= 0.0
    return offset
end